plugins/trace/trace_main.ml

```diff
@@ -55,55 +69,34 @@ module Cmdline = struct
 
   let uri = (fun s -> `Ok (uri_of_string s)), Uri.pp_hum
 
-  let dump : Uri.t option Term.t =
+  let dump : Uri.t option Config.param =
     let doc = "Dump a trace specified by $(docv)" in
-    Arg.(value & opt (some uri) None & info ["dump"] ~doc ~docv:"URI")
+    Config.(param (some uri) "dump" ~default:None ~docv:"URI" ~doc)
 
-  let load : Uri.t list Term.t =
+  let load : Uri.t list Config.param =
     let doc = "Load trace from the specified $(docv). The option maybe
     used many times to load several traces" in
-    Arg.(value & opt_all uri [] & info ["load"] ~doc ~docv:"URI")
-
-  let cmd =
-    let man = [
-      `S "SYNOPSIS";
-      `Pre "
-        $(b,bap) --$(b,$mname-dump)=$(i,URI)
-        $(b,bap) $(i,BINARY) --$(b,$mname-load)=$(i,URI)...
-       ";
-      `S "DESCRIPTION";
-      `P "Loads and prints traces. The plugin can be used in two
-       modes. When called as $(b,--$mname-dump) it will just dump the
-       specified trace and exit. In the second mode, it will load
-       specified traces, so that they can be used by
-       analysis. The loaded traces must be runs of the analyzed
-       $(i,BINARY). The loaded traces are accessible via the
-       $(b,Traces) of the traces library."
-    ] in
-    Term.(pure main $dump $load),
-    Term.info name ~doc ~man ~version:Config.version
+    Config.(param_all uri "load" ~docv:"URI" ~doc)
 
   let exitf fmt =
     kfprintf (fun ppf -> pp_print_newline ppf (); exit 1)
       err_formatter fmt
 
-  let run () =
-    match Term.eval ~catch:false ~argv cmd with
-    | `Error _ -> exit 1
-    | `Version | `Help -> exit 0
-    | `Ok (Ok `Done) -> ()
-    | `Ok (Ok `Exit) -> exit 0
-    | `Ok (Error e) -> match e with
-      | `Protocol_error err ->
-        exitf "Protocol error: %a" Error.pp err
-      | `System_error err ->
-        exitf "System error: %s" @@ Unix.error_message err
-      | `No_provider ->
-        exitf "No provider for the given URI"
-      | `Ambiguous_uri ->
-        exitf "More than one provider for a given URI"
-      | exception Incompatibe_args ->
-        exitf "Incompatible arguments, see usage SYNOPSIS"
-
-  let () = Future.upon Plugins.loaded run
+  let () =
+    Config.when_ready (fun {Config.get=(!)} ->
+        match main !dump !load with
+        | Ok `Done -> ()
+        | Ok `Exit -> exit 0
+        | Error e -> match e with
+          | `Protocol_error err ->
+            exitf "Protocol error: %a" Error.pp err
+          | `System_error err ->
+            exitf "System error: %s" @@ Unix.error_message err
+          | `No_provider ->
+            exitf "No provider for the given URI"
+          | `Ambiguous_uri ->
+            exitf "More than one provider for a given URI"
+          | exception Incompatibe_args ->
+            exitf "Incompatible arguments, see usage SYNOPSIS" )
+
```

Comment:
line 108 in the original version cannot be ignored. We should evaluate arguments only after all plugins are loaded.